rdr_service/dao/retention_eligible_metrics_dao.py

```diff
@@ -34,4 +55,14 @@ def upsert_all_with_session(self, session, retention_eligible_metrics_records):
                 record.id = dup_id
             session.merge(record)
             upsert_count = upsert_count + 1
+            # Batch up records for resource rebuilding.
+            rebuild_queue.append(int(record.participantId))
+            if len(rebuild_queue) >= 1000:
+                session.commit()
+                self._submit_rebuild_task(rebuild_queue)
+                rebuild_queue.clear()
+        # Submit any left over records for rebuilding.
+        if rebuild_queue:
+            session.commit()
+            self._submit_rebuild_task(rebuild_queue)
```

Comment:
Batching for storing the retention eligible already happens where the DAO code is called (https://github.com/all-of-us/raw-data-repository/blob/devel/rdr_service/offline/retention_eligible_import.py#L35) and uses the same batch size. Would it potentially make sense to rebuild the batches there instead of in the DAO code?

That change would also align more with the biobank_samples_pipeline, which rebuilds outside of the DAO code rather than in it (https://github.com/all-of-us/raw-data-repository/blob/devel/rdr_service/offline/biobank_samples_pipeline.py#L94)

Reply:
It's better leaving the rebuild there. The code there knows which records are going to be inserted/updated, so we don't have to rebuild all records.